File: web/src/controller.ts
// prettier-ignore
enum Button {
  A      = 0b1000_0000,
  B      = 0b0100_0000,
  Select = 0b0010_0000,
  Start  = 0b0001_0000,
  Up     = 0b0000_1000,
  Down   = 0b0000_0100,
  Left   = 0b0000_0010,
  Right  = 0b0000_0001,
}

type ButtonMap = Record<string, Button | undefined>;

class Controller {
  private value: number;
  private mappings: ButtonMap;

  constructor(mappings: ButtonMap) {
    this.value = 0b0000_0000;
    this.mappings = mappings;
  }

  handleKeyEvent(event: KeyboardEvent, state: boolean) {
    const button = this.mappings[event.code];

    if (button) {
      this.updateButton(button, state);
    }

    return button;
  }

  updateButton(button: Button, state: boolean) {
    if (state) {
      this.value |= button;
    } else {
      this.value &= ~button;
    }
  }

  state() {
    return this.value;
  }

  static playerOne() {
    return new Controller({
      KeyA: Button.A,
      KeyZ: Button.B,
      KeyQ: Button.A,
      KeyW: Button.B,
      Space: Button.Select,
      Enter: Button.Start,
      ArrowUp: Button.Up,
      ArrowDown: Button.Down,
      ArrowLeft: Button.Left,
      ArrowRight: Button.Right,
    });
  }
}

export { Controller };

File: android/app/src/main/java/dev/luckasranarison/mes/lib/Controller.kt
package dev.luckasranarison.mes.lib

import kotlin.experimental.and
import kotlin.experimental.inv
import kotlin.experimental.or

class Controller(private var value: Byte = 0b0000_0000) {
    fun update(button: Button, state: Boolean): Controller {
        val bits = (1 shl button.ordinal).toByte()
        val value = if (state) value or bits else value and bits.inv()
        return Controller(value)
    }

    fun state(): Byte = value
}

enum class Button { Right, Left, Down, Up, Start, Select, B, A }